WHMCS module

olska opened the pull request:
Start for a provisioning module for WHMCS for use by Jetpack hosting partners to provision and cancel Jetpack plans.

Currently handles 2 WHMCS core module functions for creating and terminating accounts (provisioning and cancelling Jetpack plans).

Fixes #

The PR includes a new partner directory that contains a WHMCS module to allow Jetpack hosting partners to provision Jetpack plans from WHMCS or automatically on checkout. No fixes or additions to Jetpack are included

*

#### Testing instructions:
To test the module a working WHMCS instance with a valid license as well as a Jetpack hosting partner account and the associated client id and client secret are necessary. Below are setup and usage instructions for testing once the previous 2 requirements are ready. Additional information including troubleshooting directions can be found in the full README in `/partner/WHMCS`

### Installation/Setup:
To install the module upload the ```/jetpack``` folder to the ```/modules/servers``` directory of your WHMCS installation. A server is not required for the module to function so the module will not be listed in the WHMCS Servers under the Setup tab for Products/Servers. Once uploaded to WHMCS the module is ready for use when making a new product in WHMCS. To setup a new product go to the Setup tab and locate Products/Services. You will first need to create a product group after which you can select the Create Product option and add a new product for Jetpack provisioning.

### Usage:
To use the module create a new product in one of your existing product groups. As part of the product creation process in the Module Settings tab select the “Jetpack by Automattic” module to get started. You will need to have an established Jetpack Hosting Partner account and your client_id and client_secret will be requested in order to create the product and use the module so have these handy when creating the new product. 

When setting up the module 4 custom fields will need to be added in order for the module to work correctly. These fields are verified whenever the module is utilized to create an account. The first 3 fields are required in order to provision a Jetpack plan. They can optionally be shown on the order form to allow your users to include these when checking out and assist with easier provisioning. This is the “Show on Order Form” option when creating custom fields within WHMCS. The last field is used to store details on the response from a provisioning attempt and should not be shown on the order form. When a user is provisioning a new plan this will typically be a URL that they will need to connect the plan you have provisioned for them to their https://wordpress.com account and get started with their Jetpack plan. Please use the field names exactly as shown below as these will be validated when the module is used for provisioning. The fields required are as follows:


| Field Name | Field Type | Required | Show on Order Form | Value |
| ------ | ------ | ------ | ------ | ------
| Site URL | Text Box | Yes | Optional | User input field
| Local User | Text Box | Yes | Optional | User input field
| Plan | Drop Down | Yes | Optional | (Free,Personal,Premium,Professional)
| jetpack_provisioning_details | Text Area | No | No | **Not a user field**

*

<!-- Add the following only if this is meant to be in changelog -->
#### Proposed changelog entry for your changes:

Commit:
WHMCS module

Start for a provisioning module for WHMCS for use by Jetpack hosting partners to provision and cancel Jetpack plans.

Currently handles 2 WHMCS core module functions for creating and terminating accounts (provisioning and cancelling Jetpack plans).

nb commented on partner/WHMCS/modules/servers/jetpack/jetpack.php:
```diff
@@ -0,0 +1,234 @@
+ */
+
+/**
+ * Jetpack Meta Data for WHMCS module.
```
Is this phpdoc for the `WHMCS` check or for `jetpack_MetaData`?

olska replied:
@nb This was for the jetpack_MetaData and not in the right spot after testing a few things. This has been updated. Thanks!

nb reviewed (changes requested):
Thanks, @olska! This looks good and tests well. Few comments:

* The PR description template looks unfinished – I think detailed testing instructions will be very useful (even if it requires one to set up their own WHMCS instance.
* The TravisCI build is failing, could we fix that?
* Which code style guide should we use – WHMCS or WordPress?
* Outside of error handling (which I know you’ve planned to tackle already), two other areas that need attention are logging and automated tests. From the WHMCS documentation and sample code it looks like they have good support for both. It's OK to leave it for another PR, just wanted to make sure it's on your radar.

Commit:
Add Readme and instructions

Adding a Readme that includes information on getting started as Jetpack Hosting, Installing the module and setting it up for usage for provisioning Jetpack plans

Commit:
Move WHMCS check

Move WHMCS check from before MetaData phpdock

Commit:
Add logging for module errors

Adding checking and logging for potential errors with module usage to be logged using the WHMCS logModuleCall. Also updated the checking for validating required fields to handle necessary exceptions around incorrectly setting up the module and surface/log these errors for partner hosts. Both the Create and Terminate WHMCS functions now surface host friendly provisionining errors where applicable.

brbrr commented:
> Instructions for setting up and testing the module can be found in the README.md in /partners/WHMCS

Just a quick note. It will be easier for reviewers to look directly into PR description for testing instructions.

ebinnion commented on partner/WHMCS/README.md:
````diff
@@ -0,0 +1,28 @@
+This documentation includes information for managing the WHMCS module. For documentation on the Jetpack Hosting Partners Program please refer to the [Jetpack Hosting Partner Docs][jetpack partner docs]. If you’d like to become a Jetpack Hosting Partner please visit https://jetpack.com/for/hosts/ for more information and to get started. In order to use the module a valid WHMCS license as well as a ...
+
+### Installation/Setup:
+To install the module upload the ```/jetpack``` folder to the ```/modules/servers``` directory of your WHMCS installation. A server is not required for the module to function so the module will not be listed in the WHMCS Servers under the Setup tab for Products/Servers. Once uploaded to WHMCS the module is ready for use and can be set up when making a new product in WHMCS.
````
When I initially ran through these steps, I attempted to symlink `/modules/servers/jetpack` to the `/jetpack/partner/WHMCS/modules/servers/jetpack` directory, but the module wouldn't show. So, I then actually copied the module directory from Jetpack to WHMCS and things worked. 

There's always a chance I symlinked things incorrectly. But, I figured I'd mention it in case that's a known issue.

olska replied:
Thanks for the feedback! I'll definitely take a look in to this and if there are potentially issues include them in the docs.

ebinnion commented on partner/WHMCS/README.md:
```diff
@@ -0,0 +1,28 @@
+
+
+[//]: # 
+[jetpack partner docs]: <https://github.com/Automattic/jetpack/tree/master/docs/partners>
```
What are these two lines for? I don't see them when I view the markdown file.

olska replied:
Reference links for the partner docs and info in getting started. Not really necessary and I'll inline these in another commit.

ebinnion replied:
Ahh, thanks for explaining. I just hadn't run across these before 😄

ebinnion commented on partner/WHMCS/README.md:
```diff
@@ -0,0 +1,28 @@
+| jetpack_provisioning_details | Text Area | No | No |
+
+As part of the the product creation for WHMCS you may use any of the 4 options under Module Settings for setting up the product. In the Clients Tab of WHMCS admin under Products/Services the module currently provides functionality for Create and Terminate for Jetpack plans. If you do not setup the product to be automatically provisioned on checkout you will be able perform either of these actions ...
+
```
> In the Clients Tab of WHMCS admin under Products/Services the module currently provides functionality for Create and Terminate for Jetpack plans. If you do not setup the product to be automatically provisioned on checkout you will be able perform either of these actions for your users.

I've set my product up so that it doesn't automatically provision.

![screen shot 2018-08-23 at 11 07 26 am](https://user-images.githubusercontent.com/1126811/44537605-da3c8400-a6c4-11e8-93f9-75c87b860289.png)

But, I don't see the ability to create and terminal Jetpack plans in the clients tab.

<img width="238" alt="screen shot 2018-08-23 at 11 06 08 am" src="https://user-images.githubusercontent.com/1126811/44537669-035d1480-a6c5-11e8-926c-e4bf592cf46f.png">

olska replied:
We chatted about this earlier today and I've updated the documentation to try to be clearer that the module is intended to manage and provision plans that are setup from a product purchase by the hosting partners client and a user checkout needs to be done first to see the product in the clients product listing. Let me know if the updates to the README would have helped or if you think further tweaking is needed 🙂

Commit:
Documentation update, bug fix for user checkout and enhancement.

Updated the documentation to better walk the hosts through the setup and expectations for the plugin.

There was a bug when checking out as a user and having the plan automatically provisioned. A new row was being inserted to store the next url and details instead of updating the existing row for the product order.  This is now an update instead of an insert.

Correctly handling situations where the provision endpoint returns true but the domain does not resolve. This will now store this in the jetpack_provisioning_details field to inform the host in place of the next_url

Commit:
Fix Readme

Fix markdown issue

nb commented on partner/WHMCS/modules/servers/jetpack/jetpack.php:
```diff
@@ -71,18 +71,32 @@ function jetpack_ConfigOptions()
+            return 'success';
+        }
+    } catch(Exception $e) {
+        logModuleCall('jetpack', __FUNCTION__, $params, $e->getMessage(), $e->getMessage());
```
Isn't there a better logging function than `logModuleCall`? From [the docs](https://developers.whmcs.com/provisioning-modules/module-logging/) it looks like the function's semantics are a bit different, not so much to log errors, but to log external API calls and requests.

I am curious, will the error be logged somewhere if we just return a non-“success” response in this function?

nb commented on partner/WHMCS/modules/servers/jetpack/jetpack.php:
```diff
@@ -71,18 +71,32 @@ function jetpack_ConfigOptions()
+        if ($response->success == true) {
+            return 'success';
+        }
+    } catch(Exception $e) {
```
Why is `Exception` the best choice here?

olska replied:
@nb based on earlier feedback about logModuleCall and discovery I've updated other areas to give feedback in the Activity Logs to Jetpack partner hosts on problems. Exceptions are also currently being caught and logModuleCall being used in the jetpack_createAccount and jetpack_terminateAccount functions only. I think catching an Exception here is the best choice to account for anything thats outside the realm of proper functioning of the module as well as the instances where exceptions are thrown were things may not go as intended specifically with the make_api_request function which does the provisioning/termination work. It's likely this should be outside the norm but since the create and terminate functions are core it seemed reasonable/necessary 🙂.

nb commented on partner/WHMCS/modules/servers/jetpack/jetpack.php:
```diff
@@ -204,31 +212,52 @@ function make_api_request($url, $auth='', $data=[])
+
+    foreach ($required_custom_fields as $field)
+        if (!isset($params['customfields'][$field])) {
+            throw new Exception('The module does not appear to be setup correctly. The required custom field '
```
What's the trade-off of using exceptions in `validate_required_fields` instead of returning errors?

olska replied:
Exceptions were previously used so they could be caught and used in logModuleCall but based on the feedback you previously provided above about better logging functions and further digging in to WHMCS I've updated validation to now return error strings which will automatically be logged in the Activity Logs if they are returned.

ebinnion commented on partner/WHMCS/modules/servers/jetpack/jetpack.php:
```diff
@@ -84,12 +84,13 @@ function jetpack_CreateAccount(array $params)
-
-        if ($response->success == true) {
             return 'success';
         }
```
What if `$response->success` is false?

olska replied:
@ebinnion good question and catch. I've updated the make_api_request to always return the response in the event of a 4xx error and this will also be returned for a 2xx. If this is false or $response->success is anything other than true this will now be returned as well which will add a log entry to the Activity Log.

ebinnion commented on partner/WHMCS/modules/servers/jetpack/jetpack.php:
```diff
@@ -0,0 +1,272 @@
+        $clean_url = str_replace('/', '::', $params['customfields']['Site URL']);
+        $url = 'https://public-api.wordpress.com/rest/v1.3/jpphp/' . $clean_url .'/partner-cancel';
+        $response = make_api_request($url, $access_token);
+        if ($response->success == true) {
```
It's possible for `$response->success` to be false when cancelling, which specifically happens when there was not a plan to cancel. For example, if a hosting partner calls the `/parter-cancel` endpoint twice, the first time `success` should be true and the second time `success` would be false.

olska replied:
@ebinnion good catch and thanks! I was previously handling anything that wasn't 2xx in the function that made the requests and didn't missed this. I've updated the function to now specifically handle where the response is a 200 and success is false

Commit:
Update exclusions for phpunit and tidying up

Updated phpunit.xml.dist for exclusions for whmcs. Updated code to meet PSR1/PSR2 requirements.

Commit:
lint exclusion

Lint exclusion for whmcs

Commit:
Update test_php_lint

Update test_php_lint to exclude jetpack whmcs module which requires PHP 5.6

Commit:
Update the validation function to return error strings.

WHMCS will automatically log strings that aren't "success" in the _createAccount and _terminateAccount functions in the activity log. Updating the validation function to no longer throw Exceptions but instead return strings with specific error messages or true otherwise to proceed with provisioning

Commit:
update make_api_request to provide more details to logActivity

The make_api_request function previously provided an error to the user that was not ideal for tracking down issues and required enabling Module Logging to catch the Exception and let the host know what went wrong.

If a siteurl is present on the request  the function now returns an error string with more information. WHMCS will append this to the activity log with a link to the Service Id it was placed from.

Also updated the remaining calls of logModuleCall to store a trace for outstanding Exceptions

Commit:
update make_api_request for non 2xx responses.

Updating the make_api_request function to handle non 2xx responses for failures with getting an access token to get these logged in the Activity Log.

Also updating the _terminateAccount function to handle situations where success is actually false with a 200

Clean up for PSR1/PSR2 on the save_provisioning_details function

olska commented on partner/WHMCS/modules/servers/jetpack/jetpack.php:
```diff
@@ -183,14 +187,18 @@ function make_api_request($url, $auth='', $data=[])
+
+    if ($http_code >= 400) {
+        logModuleCall('jetpack', __FUNCTION__, $url, $data, $response);
+        throw new Exception('Something Went wrong');
```
@nb I've updated this as well based on the logModuleCall feedback from earlier 🙂. I had added more in a later commit to let the user know to enable Module Logging as this error message wasn't great but am now returning an error string indicating what likely went wrong when making the API request and returning this as the error string so it can be logged in the Activity Logs.

Commit:
update make_api_requests to handle all 4xx errors

Always return the response if there is one and it is >=400 so they can be logged in the Activity Logs

Commit:
Fix bug with terminating accounts.

The function to terminate a Jetpack plan previously replaced all '/' with '::' which caused an issue the url included the scheme. Updated both the jetpack_CreateAccount and jetpack_TerminateAccount fucntions to strip the url before they are used in requests

Commit:
Update error messages and documentation

Update all non "success" error messages to include the 'JETPACK MODULE' prefix and updated the docs to include invormation on errors being logged in the WHMCS Activity Logs.

Commit:
Fix bug in get_access_token response and update validate_required_fields fucntion. Add tests for validating required fields

Fixed a typo in the get_accesss_token function and updating the validate_required_fields function to check the jetpack_provisioning_details field is present.

Commit:
Update REAMDE

The jetpack_provisioning_details is required for provisioning. Updating the README to indicate this.

Commit:
Update README

Update the documentation to indicate that automatic provisioning should not be used when the necessary fields are not show on the order form

Commit:
Update JetpackModuleTest

Remove incomplete test

nb commented on partner/WHMCS/modules/servers/jetpack/jetpack.php:
```diff
@@ -0,0 +1,288 @@
+ *
+ * @param array $params
+ * @return string 'success'
+ * @throws Exception
```
It doesn't look like the function is throwing `Exception` and also it can return strings different than “success”.

olska replied:
@nb I've updated the doc blocks for this function to indicate that 'success' is only expected by WHMCS when the process is complete otherwise an error string should be returned. Both the jetpack_createAccount and jetpack_terminateAccount functions will use the make_api_request function which will throw an exception if there is a curl error or the response body from the request is empty. I've refactored make_api_request a bit also and will post some updates in the P2 about this

nb commented on tests/php/partner/WHMCS/JetpackModuleTest.php:
```diff
@@ -0,0 +1,118 @@
+	 * Test that an error string is returned if the client id is missing
+	 * in the config options
+	 */
+	public function test_validate_required_fields_missing_client_id()
```
Thanks for adding those tests! I like that each specific situation is in its own test method.

nb commented on partner/WHMCS/modules/servers/jetpack/jetpack.php:
```diff
@@ -0,0 +1,288 @@
+        return 'JETPACK MODULE: There was a problem with provisioning or terminating a Jetpack Plan. 
+        The response was ' . $response;
+    } elseif (curl_error($curl)) {
+        throw new Exception('Unable to connect: ' . curl_errno($curl) . ' - ' . curl_error($curl));
```
In what cases do we return the error message as a string and in what cases do we throw an exception?

olska replied:
@nb I've refactored this and make_api_request no longer handles parsing the response and will not return any error strings. It either gets a response and returns it or throws an exception if there's a curl_error or the response is empty. Updates on this are also in the P2.

nb commented on partner/WHMCS/modules/servers/jetpack/jetpack.php:
```diff
@@ -0,0 +1,288 @@
+    $response = curl_exec($curl);
+    $http_code = curl_getinfo($curl, CURLINFO_HTTP_CODE);
+
+    if ($response && $http_code >= 400) {
```
What if the response is 301 or 100?

olska replied:
@nb great question! I understood what a successful response should be but handled edge cases inadequately including these 2 that you mentioned. This is no longer in the make_api_request function and will be handled by the calling function so there's more room for error management and clarity on what to do when specific errors occur that were legitimate responses and not curl errors.

nb commented on partner/WHMCS/modules/servers/jetpack/jetpack.php:
```diff
@@ -0,0 +1,288 @@
+ *  - Required Config Options
+ *
+ * @param array $params
+ * @return bool
```
When there is an error it returns a string.

olska replied:
@nb I've updated this to reflect the mixed return type and also updated the functions that call this to check for the boolean value or return the error string.

nb commented on partner/WHMCS/modules/servers/jetpack/jetpack.php:
```diff
@@ -0,0 +1,288 @@
+
+            return 'success';
+        } else {
+            return $response;
```
Wouldn't `$response` be an object here? Is it OK to return an object, instead of a string?

olska replied:
@nb the make_api_request function was previously trying to handle the response object and return an error string if the response wasn't 2xx which would have technically caught this. In hindsight this isn't great and actually really confusing in the state it is where you commented on things as $response could have either been an object or a string. make_api_request no longer does this and will always return an object and it's up to the calling function to handle things from there.

Commit:
Updates to make_api_request, addeded error handling for provisioning and cancellation and code cleanup

Updated the make_api_request to simply make the api request and no longer do any processing of the response. If there is a response and no curl errors the response will be json_decoded and now sent to the function that made the request.

Added 2 functions for handling errors specific to provisioning and cancelling a jetpack plan. These are intended to be able to parse the response from a request to provision or cancel including looking at the status code and the message in the response body if there is one to provide a useful error message.

Also updated the docs for most functions as they were either incomplete/incorrect or missing details about params and expectation.

Commit:
Docs update

Update docs for functions that utilize make_api_request which throws an exception if there is a curl_error or an empty response

Commit:
Update validate_required_fields docs

Update validate_required_fields to return true if there are no invalid requried fields and calling functions to verify this. Also update docs for function to reflect mixed return type.

Commit:
update jetpack_TerminateAccount

Update the jetpack terminate account function to strip trailing slashes before replacing other slashes with :: to make the cancellation request.

olska commented on partner/WHMCS/modules/servers/jetpack/jetpack.php:
```diff
@@ -0,0 +1,288 @@
+    try {
+        $access_token = get_access_token($params);
+        $stripped_url = preg_replace("(^https?://)", "", $params['customfields']['Site URL']);
+        $clean_url = str_replace('/', '::', $stripped_url);
```
@ebinnion it is very possible and the url would look something like verysadurlgoeshere.com::. Thanks for this! I've updated the _TerminateAccount to first strip trailing slashes and then the url scheme before replacing other slashes.

olska commented on partner/WHMCS/modules/servers/jetpack/jetpack.php:
```diff
@@ -0,0 +1,288 @@
+            return 'success';
+        } elseif ($response->success === false) {
+            return 'JETPACK MODULE: Unable to terminate this Jetpack plan as it has likely already been cancelled';
+        }
```
@ebinnion thanks for insight on this. I think ultimately make_api_request which has been updated was doing too much and getting an access token needed to be identified as a point of failure that should be addressed before attempting either provisioning or terminating a Jetpack Plan. I've updated both the _TerminateAccount and _CreateAccount functions to handle getting an access token failure outside of their normal function to account for this.

olska replied:
Also realized I didn't answer your question specifically about what happens if an invalid token is used. The module currently doesn't cache the token so we get a new one each time. I did however add a response handler function where api requests are being made including getting an access token that will check the response whenever the expected result is not returned and make a useful error message for the partner host based on the known codes from the docs and if there is a message or error_message in the body.

Commit:
Add response handler for authentication and update termniation response handler

Added a response handler specifically for when getting an access token is not succesful. Also updated the termination response handler to look for 404's when attempting to cancel a plan.

Commit:
Update docblock

Update jetpack_CreateAccount docblock

Commit:
Docs update

Update some of the docblocks to be clearer about

Commit:
update Exception handling

Remove the try/catch from the jetpack_TerminateAccount and jetpack_CreateAccount. Exceptions will be logged to the WHMCS Activity Log.

Commit:
Adding tests and bug fix.

Add tests for error response handling and fix bug with responses handling server errors. They previously did not include handling 500 errors.

Commit:
update doc blocks

Remove @throws on functions calling make_api_requests.

ebinnion commented on partner/WHMCS/modules/servers/jetpack/jetpack.php:
```diff
@@ -251,8 +251,9 @@ function make_api_request($url, $auth = null, $data = null)
 /**
- * Save the next_url for Jetpack activation/setup to the
- * order for the client
+ * Save the next_url for Jetpack activation/setup to the order for the client. If a next url is not
```
> If a next url is not and the pending param

Perhaps this should be "If a next url is not present" or something similar?

ebinnion commented on partner/WHMCS/modules/servers/jetpack/jetpack.php:
```diff
@@ -0,0 +1,365 @@
+ * Pre Provisioning Steps:
+ *  Module requirements are validated and will return an error string
+ *  if the module was not setup correctly. An error string will also be
+ *  returned in the even that a request to provision a new returns a
```
"in the even" => "in the event"

olska replied:
Fixed this and reviewing all the doc blocks again.

ebinnion commented on partner/WHMCS/modules/servers/jetpack/jetpack.php:
```diff
@@ -0,0 +1,365 @@
+ *  returned in the even that a request to provision a new returns a
+ *  4xx or 5xx error.
+ *
+ *  An Access token will also be retrieved before trying to provisioning.
```
"before trying to provisioning." => "before trying to provision."

olska replied:
Also fixed. Thanks!

ebinnion commented on partner/WHMCS/modules/servers/jetpack/jetpack.php:
```diff
@@ -0,0 +1,365 @@
+    ];
+
+    $response = make_api_request($provisioning_url, $access_token, $request_data);
+    if ($response->success && $response->success == true) {
```
What happens if `$response` does not have a `success attribute?

olska replied:
This was falling back to the final else and returning an error string. I have added a check to make sure that it's set before attempting to access it.

ebinnion commented on partner/WHMCS/modules/servers/jetpack/jetpack.php:
```diff
@@ -0,0 +1,365 @@
+    } else {
+        $errors = get_provisioning_errors_from_response($response);
+        return $errors;
+    }
```
I like the way this conditional is setup. If I were to write it, I think I'd do:

```
if ( ! empty() ) {
  // stuff
} else {
  // stuff
}
```  

But, it reads well and we know that `success` is getting returned when expected now. 👍

olska replied:
I made a small change to this to help make things less confusing and removed an unneeded check for pending accounts.

ebinnion commented on partner/WHMCS/modules/servers/jetpack/jetpack.php:
```diff
@@ -0,0 +1,365 @@
+
+    $request_url = 'https://public-api.wordpress.com/rest/v1.3/jpphp/' . $clean_url . '/partner-cancel';
+    $response = make_api_request($request_url, $access_token);
+    if ($response->success === true) {
```
What happens if the `success` attribute of `$response` is not set?

olska replied:
This was similar to the jetpack_CreateAccount. I've added a check on this one as well to make sure the property is set before continuing.

ebinnion commented on partner/WHMCS/modules/servers/jetpack/jetpack.php:
```diff
@@ -0,0 +1,365 @@
+ *
+ *
+ * @param array $params WHMCS params
+ * @return mixed A string with the access token or an error string beginning with 'JETPACK MODULE' indicating that
```
Should this be `@return string` now that we're returning a string in both cases?

olska replied:
I've updated this also based on another comment to return an array which has either the token or an error set

ebinnion commented on partner/WHMCS/modules/servers/jetpack/jetpack.php:
```diff
@@ -0,0 +1,365 @@
+ * Include the http status in the response.
+ *
+ * @param string $url where to make the request to
+ * @param null $auth access token to make a provisioning or cancellation request
```
Is `@param null` correct? It looks like `$auth` is expected to be a string.

olska replied:
Updated these to be the expected values when they are passed

ebinnion commented on partner/WHMCS/modules/servers/jetpack/jetpack.php:
```diff
@@ -0,0 +1,365 @@
+
+
+    $decoded_response = json_decode($response);
+    $decoded_response->http_status = $http_code;
```
Will this result in us returning the HTTP status code mixed in with the response from the API?

olska replied:
@ebinnion This will. It's included in some messages already as part of an error message but not all of them.

olska replied:
@ebinnion thanks for the suggestion on this! I looked in to the documentation and http_envelope argument will accomplish the same thing in most but not all calls for the module. This means that I'd have to check the response or the url query params and set it in some cases. I think for now setting it with ```curl_getinfo``` works but for a future version if there's need to refactor make_api_request to change how requests are handled it would be worth looking in to using http_envelope when it's available 🙂.

ebinnion commented on partner/WHMCS/modules/servers/jetpack/jetpack.php:
```diff
@@ -0,0 +1,365 @@
+
+/**
+ * Save the next_url for Jetpack activation/setup to the order for the client. If a next url is not
+ * and the pending param is true save a message letting the user know a plan is pending and will
```
"If a next url is not and" should probably be something like, "If a next url is not present or is empty and"

olska replied:
Also fixed.

ebinnion commented on partner/WHMCS/modules/servers/jetpack/jetpack.php:
```diff
@@ -0,0 +1,365 @@
+        return $response->access_token;
+    } else {
+        $errors = get_authentication_errors_from_response($response);
+        return $errors;
```
Are there any pros/cons to if we returned `false` here?

olska replied:
This one was on my mind also as the check for verifying if we got a real access token or an error seemed less than ideal. I thought about returning false however I also wanted to be able to get the error message to be able to provide this to the host and keep things together. I've updated the function to return an array that either has the access token set or an error set which can be checked for by calling functions and the access token set for subsequent calls or the error string returned

ebinnion commented:
Hi @olska – I've looked over the code and left a few notes. I like that way things are shaping up! 

There are several typos and grammar issues that in docblocks that I'd suggest watching for going forward. Besides that, I pointed out a couple of small potential issues as well as some small details.

At this point, I have not functionally tested yet.

Commit:
Update documentation

Fix issues where sentences weren't complete or had some gramattical errors. Also updated the docs for get_authentication_errors_from_response to be more clear

olska commented on partner/WHMCS/modules/servers/jetpack/jetpack.php:
```diff
@@ -0,0 +1,365 @@
+/**
+ * If we are attempting to get an access token and this fails parse the response
+ * http status code and response body and return a useful error from these
+ * regarding what went wrong.
```
I've split this up a bit and explained the specific check thats happening here and whats being done to make the error string thats being returned.

Commit:
update jetpack_CreateAccount and jetpack_TerminateAccount

Update the create account and terminate account functinos to check that the response object has the success property set before use

Commit:
remove unecessary check

if the next url is not set in the response for the jetpack_createAccount function the plan provisioning is pending

Commit:
update make_api_request

fix params on dock block for make_api_request function

Commit:
update get_access_token return value

Update get access token return value to either include an access token or errors that occurred while trying to get one

olska commented:
@ebinnion many thanks for the notes! I've made more changes based on your comments and also some ideas I had as well around simplifying things. I did have one question on one comment which I left for you in regards to handling termination errors. Thanks!

ebinnion commented on partner/WHMCS/modules/servers/jetpack/jetpack.php:
```diff
@@ -0,0 +1,365 @@
+    } elseif ($response->http_status >= 500) {
+        return 'JETPACK MODULE: There was an error communicating with the provisioning server. Please try again later.';
+    }
+    return 'JETPACK MODULE: There was an error provisioning the Jetpack plan. Please contact us for assistance.';
```
Interesting. Can you expand on which cases the error message is in `$response->message` and in which case it's in `$response->error_message`?

olska replied:
@ebinnion after testing this again I've actually only ever gotten $response->message and never $response->error_message testing all the scenarios I could in the partner docs for provisioning. I did go ahead and update the cancellation function to also include the $response->message if there is one. Is it possible that the partner docs need to be updated to ```message``` instead of ```error_message``` for provisioning and cancelling endpoints or are there other use cases where it may be in the response?

ebinnion commented on partner/WHMCS/modules/servers/jetpack/jetpack.php:
```diff
@@ -0,0 +1,376 @@
+ *
+ * @param array WHMCS $params
+ * @return string Either 'success' or an error with what went wrong when provisioning
+ */
```
I like this docblock. 👍

ebinnion reviewed (commented):
Hi @olska – This is looking good. I left a couple of comments. I was able to successfully provision and cancel a plan via your WHMCS install.

I did not work on error states though.

ebinnion commented:
<img width="883" alt="screen shot 2018-09-07 at 4 34 48 pm" src="https://user-images.githubusercontent.com/1126811/45244560-fee84c80-b2bd-11e8-8a96-c1173c65581f.png">

It's likely out of the scope of this project, but do you have any ideas on how we can simplify choosing the local user? Is it possible to query for users and display a list of admins or something similar?

Commit:
Clean up error handling and update jetpack_CreateAccount

If there is a $response->message always return it in some context regardless of the https status code. Also update jetpack_CreateAccount to look for the boolean true value on the $response->success

Commit:
Clean up errors

Clean up error strings when the response has a message.

nb commented on partner/WHMCS/modules/servers/jetpack/jetpack.php:
```diff
@@ -0,0 +1,383 @@
+
+    if (!in_array(strtolower($params['customfields']['Plan']), $allowed_plans)) {
+        return 'JETPACK MODULE: The module does not appear to be setup correctly. ' .
+            $params['customfields']['Plan'] . ' is not an allowed plan';
```
This value is used in the WHCMS special functions, do we know whether they are used in HTML context and do we need to escape the user-supplied values here?

nb replied:
@olska my question is if `$params['customfields']['Plan']` is `<script>alert('hi!');</script>`, would it be executed on a HTML page somewhere?

olska replied:
@nb both the user and host supplied custom field values are escaped already when saved so this alert will not be executed on any HTML pages anywhere. We do not need to escape these in the context of the module which currently uses core WHMCS functions however if anything custom is done revisiting this would be necessary.

olska replied:
@nb I also did verify this in the database table for custom fields to be sure things are as expected.

nb replied:
@olska maybe I am missing something, I am sorry for the repeated questions, but where exactly is the value that a user enters in the Plan field (even if it's a drop-down, it's trivial for the user to send any value they want) escaped before saving? Is it somewhere inside WHMCS? Or somewhere in our code that I have missed?

olska replied:
@nb no worries about the questions and thanks for asking for clarification. It's a legitimate concern so I do want to make sure we are on the same understanding and if there's a change that needs to be made to resolve a concern it is addressed. The custom fields are not created or managed by the module. The host sets these up in the WHMCS product settings as part of the set up steps. A user can send any value they want but these are managed and saved by WHMCS and escaped accordingly. WHMCS code is specifically handling this right now and the values are escaped from wherever they are created including on the order page when the user sends their details to the host to get a Jetpack plan.

ebinnion commented:
I'm going to close this PR as we've moved the functionality to https://github.com/Automattic/Jetpack-Partners-WHMCS-Module
